drivers: adc: stm32: Disable ADC before calibration

pdietl opened the pull request:
The STM32 docs state that the ADC may not be calibrated unless
the ADC is disabled (ADEN=0). This commit implements this constraint

Fixes #40936

Signed-off-by: Pete Dietl <petedietl@gmail.com>

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -592,7 +592,12 @@ static int start_read(const struct device *dev,
 	!defined(CONFIG_SOC_SERIES_STM32F1X) && \
 	!defined(STM32F3X_ADC_V2_5) && \
 	!defined(CONFIG_SOC_SERIES_STM32L1X)
+		LL_ADC_Disable(adc);
```
Few lines above, the ADC has just been enabled after setting the oversampling bits, for the STM32L0x, but it seems that all soc requires ADSTART =0. 
Could we find and common combination to avoid disabling/waiting/enabling/disabling/re-enabling ?

pdietl replied:
That is true for the G4 at least. Should we check the data sheets for all other STM32s? It seems like we can combine disabling the ADC for oversampling calibration setting.

![image](https://user-images.githubusercontent.com/10335942/145010750-5ba77425-51ea-4919-bd87-a05eb4243d97.png)

pdietl replied:
@FRASTM in regards to your comment in the issue, in  RM0008, it recommends calibrating once after power-on, so I suppose that is why requesting calibration before starting an ADC read is not supported in the code. Do you concur?

mbolivar-nordic reviewed (commented):
@erwango I'm not sure what review you are requesting from me -- this looks like a pure SoC internal implementation change. But no objections.

erwango commented:
> @erwango I'm not sure what review you are requesting from me -- this looks like a pure SoC internal implementation change. But no objections.

Followed a gh suggestion blindly. But indeed, I'm not sure what was the rationale. Let me remove you ;-)

pdietl commented:
@FRASTM Just poking to make sure you saw my reply to your comment. I think fulfil your suggestion, if we assume that we don't want to allow additional calibration of an ADC after the initial calibration for the STM32F1 (as is currently implemented in the code), then I think I should be able to perform the setting of the oversampling bits and perform a calibration while the ADC is disabled, and then re-enable it after both are finished. Is this a better solution?

FRASTM reviewed (approved):
@pdietl as a second step PR,  your proposal to group is welcome

erwango reviewed (changes requested):
Nack'ed to avoid merge before I can review

pdietl commented:
> @pdietl as a second step PR, your proposal to group is welcome

I misunderstood and did it all at once 😅

pdietl commented:
@FRASTM poke.

pdietl commented:
@erwango can you review please :)

erwango reviewed (changes requested):
@pdietl I'm sorry but this change is confusing as not exactly fitting the description.
I agree that current driver implementation doesn't help to make things clear, and this is what brings me to be extra careful on the claims and the changes.

pdietl commented:
@erwango Should I remove the changes of combining the previous block with adc disabling? I can add that as another PR.

erwango commented:
> @erwango Should I remove the changes of combining the previous block with adc disabling? I can add that as another PR.

This would be a good start. Thanks

pdietl committed:
drivers: adc: stm32: Disable ADC before calibration

The STM32 docs state that the ADC may not be calibrated unless
the ADC is disabled (ADEN=0). This commit implements this constraint

Fixes #40936

Signed-off-by: Pete Dietl <petedietl@gmail.com>

pdietl commented:
@erwango done

pdietl commented:
Who can merge?

erwango commented:
> Who can merge?

Need two +1 before merge. I asked @FRASTM to review again due to initial code change, so right now you still miss one +1
@FRASTM would you mind reviewing this once again?

pdietl commented:
poke @FRASTM
